### PR TITLE
Cortex-M: Use q/dq ops in Arm Ethos Runner

### PR DESCRIPTION
**Pull request by digantdesai**
### Summary
* Adds MVE version of [de]quantize_per_tensor
* Adds basic support to link Cortex-M backend op (replacing ET quantized ops) with Arm Ethos runner
* Use them in FVP tests for M55/M85
* No selective build yet, the plan is to extend this such that we can have, merged_yaml with ops used my model(s) from Cortex-M-backend and Portable Lib.

### Test plan
```./examples/arm/run.sh --target=ethos-u85-256 --model_name=lstm --bundleio```

### CODE REVIEW

File: examples/arm/aot_arm_compiler.py [hunk 1]
```diff
@@ -759,7 +840,9 @@ def to_edge_no_delegate(exported_program, args, model: torch.nn.Module, example_
             output_name = os.path.join(args.output, output_name)
 
     if args.bundleio:
-        save_bpte_program(exec_prog, original_model, output_name)
+        # Realize the quantization impact on numerics when generating reference output
```
Comment (AdrianLundell) [inline examples/arm/aot_arm_compiler.py:843]: The problematic thing with comparing to the quantized model is that you don't catch how changes to the quantization affect accuracy compared to the original. Is this tested somewhere else?

File: backends/cortex_m/ops/op_dequantize_per_tensor.cpp [hunk 2]
```diff
@@ -106,29 +111,71 @@ Tensor& dequantize_per_tensor_out(
...
 #if defined(HAS_HELIUM_SIMD)
-// Helium MVE implementation for float32 to int8 quantization
-#Error "Implement MVE version!"
-#else
-  // Scalar implementation for float32 to int8 quantization
-  for (size_t i = 0; i < numel; i++) {
-    out_data[i] =
-        dequantize_val<int8_t, float>(scale, zp, input_data[i], qmin, qmax);
+  // Helium MVE implementation for int8 to float quantization
+  static uint8x16_t voffset{
```
Comment (AdrianLundell) [inline backends/cortex_m/ops/op_dequantize_per_tensor.cpp:126]: These implementations are actually very similar to what we have done on our side except that we used predication for the tail handling, is this something you have considered?

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp [hunk 3]
```diff
@@ -75,18 +75,18 @@ void check_quantize_args(
 /**
  * Scalar implementation of quantization for a single value.
  */
-template <typename T, typename K>
-T quantize_val(
-    float inv_scale,
+template <typename Q, typename F>
```
Comment (AdrianLundell) [inline backends/cortex_m/ops/op_quantize_per_tensor.cpp:78]: Why template this when only int8/f32 is supported?

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp [hunk 4]
```diff
@@ -123,16 +123,97 @@ Tensor& quantize_per_tensor_out(
...
+      0x6,
+      0xE,
+      0x3,
+      0xB,
+      0x7,
+      0xF};
+
+  float32x4_t inv_scale_vec = vdupq_n_f32(inv_scale);
+
+  // Magic number for float to int conversion, round to nearest even integer
```
Comment (AdrianLundell) [inline backends/cortex_m/ops/op_quantize_per_tensor.cpp:150]: Really cool trick, I didn't know about this. This looks better than the rounding I have suggested.

File: examples/arm/executor_runner/CMakeLists.txt [hunk 5]
```diff
@@ -505,7 +504,18 @@ set_property(
   PROPERTY IMPORTED_LOCATION
            "${ET_BUILD_DIR_PATH}/kernels/quantized/libquantized_kernels.a"
 )
-
+add_library(cortex_m_ops_lib STATIC IMPORTED)
```
Comment (AdrianLundell) [inline examples/arm/executor_runner/CMakeLists.txt:507]: Would be nice with a comment here to explain the discrepancy to how the kernel registration library is used here compared to how it is described here https://docs.pytorch.org/executorch/stable/kernel-library-custom-aten-kernel.html#yaml-entry-api-high-level-architecture

File: examples/arm/aot_arm_compiler.py (hunk 1 above)
Comment (digantdesai) [reply examples/arm/aot_arm_compiler.py:843]: You are right, IMO that should be tested independently?

A - original FP32 model
B - quantized model
C - lowered model

(1) A vs. B captures quantization noise
(2) B vs. C captures implementation noise

A vs. C isn't really useful in case of quantization esp when you are trying vet the implementation, which is the goal here with bundled-io to test the implementation on FVP, while (1) can be done using random op impls AoT or on x86.

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp (hunk 3 above)
Comment (digantdesai) [reply backends/cortex_m/ops/op_quantize_per_tensor.cpp:78]: Just copy-paste artifact, it shouldn't cause any size or perf overhead.

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp (hunk 4 above)
Comment (digantdesai) [reply backends/cortex_m/ops/op_quantize_per_tensor.cpp:150]: Thanks. This isn't too uncommon for f32 -> i32.

File: examples/arm/executor_runner/CMakeLists.txt (hunk 5 above)
Comment (digantdesai) [reply examples/arm/executor_runner/CMakeLists.txt:507]: Yeah this isn't doing any selection yet. I left a comment on the top level PR.

File: examples/arm/aot_arm_compiler.py (hunk 1 above)
Comment (AdrianLundell) [reply examples/arm/aot_arm_compiler.py:843]: Agreed, I don't think we need to change something here as long as we keep this in mind because I feel like we are lacking somewhat in (1) right now afaik.

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp (hunk 3 above)
Comment (AdrianLundell) [reply backends/cortex_m/ops/op_quantize_per_tensor.cpp:78]: Sure, I'm just thinking it could cause confusion about what is supported but no strong opinion here

File: examples/arm/executor_runner/CMakeLists.txt (hunk 5 above)
Comment (AdrianLundell) [reply examples/arm/executor_runner/CMakeLists.txt:507]: Ack

File: backends/cortex_m/ops/op_quantize_per_tensor.cpp (hunk 4 above)
Comment (AdrianLundell) [reply backends/cortex_m/ops/op_quantize_per_tensor.cpp:150]: I guess in CMSIS-NN everything was integer already so I didn't have to worry about that, but nice to learn.

**Review by AdrianLundell — APPROVED** [review]

**Comment by digantdesai** [comment]
@zingo - `test_sigmoid_u85_BI_add_sigmoid` failure is related or flaky? I can try to reproduce locally.

**Comment by zingo** [comment]
> @zingo - `test_sigmoid_u85_BI_add_sigmoid` failure is related or flaky? I can try to reproduce locally.
> 
> Update: Landing this given I don't think we are using these q/dq ops when lowering through tester, which I can do in the follow up PRs. Also U55 tests are passing.

There is a PR where we hope to have fixed the flakey sigmoid, and update vela that you brought into Meta for test/merge a few days ago. It will hopefully fix the flakeyness. Until it is merged i think rerun the test is the best option (or ignore it)

**Comment by zingo** [comment]
This is the sigmoid fix PR mentioned: https://github.com/pytorch/executorch/pull/10548
